File: leetcode/200_number_of_islands/200.py
class Solution:
    def bfs(self, grid, x, y, visited):
        queue = [(x, y)]
        
        while queue:
            x, y = queue.pop(0)
            
            for (dx, dy) in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                x2 = x + dx
                y2 = y + dy
                
                if (
                    (0 <= x2 < len(grid[0]))
                    and (0 <= y2 < len(grid))
                    and (not visited[y2][x2])
                ):
                    visited[y2][x2] = True
                    
                    if grid[y2][x2] == "1":
                        queue.append((x2, y2))
    
    def numIslands(self, grid: List[List[str]]) -> int:
        nrows = len(grid)
        ncols = len(grid[0])
        visited = [[False for _ in range(ncols)] for _ in range(nrows)]
        
        num_islands = 0
        
        for y in range(nrows):
            for x in range(ncols):
                if not visited[y][x]:
                    visited[y][x] = True
                    
                    if grid[y][x] == "1":
                        self.bfs(grid, x, y, visited)
                        num_islands += 1
                        
        return num_islands

File: leetcode/200_number_of_islands/200.cpp
class Solution {
public:
    void bfs(vector<vector<char>>& grid, int i, int j) {
        queue<pair<int, int>> q;
        q.emplace(j, i);
        grid[i][j] = '2';
        
        vector<pair<int, int>> directions {
            make_pair(1, 0),
            make_pair(0, 1),
            make_pair(-1, 0),
            make_pair(0, -1)
        };
        
        while (!q.empty()) {
            auto node {q.front()};
            q.pop();
            
            for (pair<int, int> direction : directions) {
                int new_x {node.first + direction.first};
                int new_y {node.second + direction.second};
                
                if (
                    new_x >= 0 && new_x < grid[0].size()
                    && new_y >= 0 && new_y < grid.size()
                    && grid[new_y][new_x] == '1'
                ) {
                    q.emplace(new_x, new_y);
                    grid[new_y][new_x] = '2';
                }
            }
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int num_islands {0};
        
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == '1') {
                    bfs(grid, i, j);
                    num_islands++;
                } else {
                    grid[i][j] = '2';
                }
            }
        }
        return num_islands;
    }
};
